cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -405,155 +406,149 @@ runQueryKesPeriodInfo (AnyConsensusModeParams cModeParams) network nodeOpCertFil
+
       -- We get the operational certificate counter from the protocol state and check that
       -- it is equivalent to what we have on disk.
 
       let ptclStateQinMode = QueryInEra eInMode . QueryInShelleyBasedEra sbe $ QueryProtocolState
       ptclState <- executeQuery era cModeParams localNodeConnInfo ptclStateQinMode
-      (opCertCounter, ptclStateCounter, opCertCounterStateDiag) <- opCertCounterStateCheck ptclState opCert
-
-      let diagnoses = opCertCounterStateDiag ++ [periodCheckDiag]
-
-      if any anyFailureDiagnostic diagnoses
-      then
-        case mOutFile of
-          Just (OutputFile _) -> liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
-          Nothing -> liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
-      else do
-        let kesPeriodInfo = O.QueryKesPeriodInfoOutput
-                              { O.qKesInfoCurrentKESPeriod = currentKesPeriod
-                              , O.qKesInfoStartKesInterval = kesIntervalStart
-                              , O.qKesInfoStartEndInterval = kesIntervalEnd
-                              , O.qKesInfoRemainingSlotsInPeriod = slotsTillNewKesKey
-                              , O.qKesInfoNodeStateOperationalCertNo = ptclStateCounter
-                              , O.qKesInfoOnDiskOperationalCertNo = opCertCounter
-                              , O.qKesInfoMaxKesKeyEvolutions = fromIntegral protocolParamMaxKESEvolutions
-                              , O.qKesInfoSlotsPerKesPeriod = fromIntegral protocolParamSlotsPerKESPeriod
-                              , O.qKesInfoKesKeyExpiry = kesKeyExpirySlotUtc
-                              }
-            kesPeriodInfoJSON = encodePretty kesPeriodInfo
-
-        liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
-
-        case mOutFile of
-          Just (OutputFile oFp) ->
-            handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError oFp)
-              $ LBS.writeFile oFp kesPeriodInfoJSON
-          Nothing -> liftIO $ LBS.putStrLn kesPeriodInfoJSON
+
+      (onDiskC, stateC) <- opCertOnDiskAndStateCounters ptclState opCert
+      let counterInformation = opCertNodeAndOnDiskCounters onDiskC stateC
+
+      -- Always render diagnostic information
+      liftIO . putStrLn $ renderOpCertIntervalInformation nodeOpCertFile opCertIntervalInformation
+      liftIO $ renderOpCertNodeAndOnDiskCounterInformation nodeOpCertFile counterInformation
+
+      let qKesInfoOutput = createQueryKesPeriodInfoOutput opCertIntervalInformation counterInformation eInfo gParams
+          kesPeriodInfoJSON = encodePretty qKesInfoOutput
+      case mOutFile of
+        Just (OutputFile oFp) -> do
+          liftIO $ LBS.putStrLn kesPeriodInfoJSON
+          handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError oFp)
+            $ LBS.writeFile oFp kesPeriodInfoJSON
+        Nothing -> liftIO $ LBS.putStrLn kesPeriodInfoJSON
```

Comment:
To prevent duplication, suggest:

```
      liftIO $ LBS.putStrLn kesPeriodInfoJSON
      forM_ mOutFile $ \Just (OutputFile oFp) -> do
          handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError oFp)
```